### ISSUE TITLE
The masked mean, which is the focal point of the paper, is not utilized at all.

### ISSUE THREAD

**Issue by linhlt-it-ee** [andyrdt/refusal_direction#10]
Hello! The paper is intriguing. However i figured out an issue. The masked_mean is invoked only once within the kl_div_fn, which itself is called just once with the following expression:
```
kl_div_fn(baseline_harmless_logits, intervention_logits, mask=None).mean(dim=0).item()
```
Then when is masked_mean used?

**Comment by andyrdt** [comment]
I agree it looks like `masked_mean` is not called, and it seems like we forgot to clean this up in our code. But I don't think there's a bug.

The `kl_div_fn` is computing the KL divergence between two tensors of logits. If the mask is none, it will compute the KL divergence between these two tensors of logits normally. If mask is passed in, it will only compute the KL divergence of logits at positions specified by the mask. When implementing this we were probably playing around with batches that varied in length with padding tokens, and so we would have wanted to specify not to include padding token positions for the KL divergence computation. But the way things are written in the codebase now, we're only computing KL divergence on the last token position of each sequence (you can see this in `get_last_position_logits()`). Since we're only looking at a single position, there's no need for masking (all batch sequences are the same length) - that's why `kl_div_fn` is always called with `mask=None`.

It's also not correct that this masked_mean function is "the focal point of the paper" - the `kl_div_fn` is just used to compute the "kl_score" for direction selection (see [appendix C](https://arxiv.org/pdf/2406.11717)). The computation of the difference-in-means vector is in `generate_directions.py`.
